### ISSUE TITLE
Some clarifications and potential bug?

### ISSUE THREAD

**Issue by apoorv2904** [lucidrains/audiolm-pytorch#48]
HI, first of all thank you for putting the efforts to reproduce Soundstream code. I have a few questions to which I would greatly appreciate any clarifications.

1. I notice that you use encoder_attn and decoder_attn with LocalMHA. But I couldn't find where the attention parameters are being optimized. If I understood correctly the function: **non_discr_parameters** only returns encoder and decoder parameters and not the attention. Please  correct me if I missed something.

2. For the SoundStreamTrainer, train_step, shouldn't we **first set the dicriminator grads to zero** before calling the backward on the discriminator losses?  The generator losses would have also added gradients to the discriminator parameters. Currently, the discriminator losses backward pass will add to those if I understood correctly. Again please let me know if I misunderstood something.

Thanks again,
Apoorv

**Comment by lucidrains** [comment]
@apoorv2904 omg, thank you Apoorv! :pray: i may have this discriminator bug in a number of other repos too :man_facepalming:

**Comment by apoorv2904** [comment]
Thank you for the quick responses and fixes as well as the library 🙏. I will close the issue now.

**Comment by apoorv2904** [comment]
@lucidrains I tried training Soundstream model following the readme as shown below. However the validation samples seem to be only noise even after 10K steps. Were you or any one else able to train to get the SoundStream training to work? Would you have any comment on what I might be doing wrong?  Would it be possible to share any training graph? FYI,  I am now trying 25K steps with `data_max_length = 320 * 75` but that doesn't seem to work for now.

Thanks a lot! 🙏

```
from audiolm_pytorch import SoundStream, SoundStreamTrainer

soundstream = SoundStream(
    codebook_size = 1024,
    rq_num_quantizers = 8,
)

trainer = SoundStreamTrainer(
    soundstream,
    folder = '/path/to/librispeech',
    batch_size = 4,
    grad_accum_every = 8,         # effective batch size of 32
    data_max_length = 320 * 32,
    num_train_steps = 10000
).cuda()

trainer.train()
```
